tests/test_unit/test_auth/test_auth_keycloak.py

```diff
@@ -0,0 +1,147 @@
+import time
+
+import pytest
+import responses
+from httpx import AsyncClient
+
+from syncmaster.backend.settings import BackendSettings as Settings
+from tests.mocks import MockUser
+from tests.test_unit.test_auth.mocks.keycloak import (
+    create_session_cookie,
+    mock_keycloak_realm,
+    mock_keycloak_well_known,
+)
+
+KEYCLOAK_PROVIDER = "syncmaster.backend.providers.auth.keycloak_provider.KeycloakAuthProvider"
+pytestmark = [pytest.mark.asyncio, pytest.mark.backend]
+
+
+@responses.activate
+@pytest.mark.parametrize("settings", [{"auth": {"provider": KEYCLOAK_PROVIDER}}], indirect=True)
+async def test_get_keycloak_user_unauthorized(client: AsyncClient):
+    mock_keycloak_well_known(responses)
+
+    response = await client.get("/v1/users/some_user_id")
+
+    # redirect unauthorized user to Keycloak
+    assert response.status_code == 307
+    assert "protocol/openid-connect/auth?" in str(
+        response.next_request.url,
+    )
+
+
+@responses.activate
+@pytest.mark.parametrize("settings", [{"auth": {"provider": KEYCLOAK_PROVIDER}}], indirect=True)
+async def test_get_keycloak_user_authorized(
+    client: AsyncClient,
+    simple_user: MockUser,
+    settings: Settings,
+    access_token_factory,
+):
+    payload = {
+        "sub": str(simple_user.id),
+        "preferred_username": simple_user.username,
+        "email": simple_user.email,
+        "given_name": simple_user.first_name,
+        "middle_name": simple_user.middle_name,
+        "family_name": simple_user.last_name,
+        "exp": time.time() + 1000,
+    }
+
+    mock_keycloak_well_known(responses)
+    mock_keycloak_realm(responses)
+
+    session_cookie = create_session_cookie(payload, settings.server.session.secret_key)
+    headers = {
+        "Cookie": f"session={session_cookie}",
+    }
+    response = await client.get(
+        f"/v1/users/{simple_user.id}",
+        headers=headers,
+    )
+
+    assert response.status_code == 200
+    assert response.json() == {
+        "id": simple_user.id,
+        "is_superuser": simple_user.is_superuser,
+        "username": simple_user.username,
+    }
+
+
+@responses.activate
+@pytest.mark.parametrize("settings", [{"auth": {"provider": KEYCLOAK_PROVIDER}}], indirect=True)
+async def test_get_keycloak_deleted_user(
```

Comment:
Another case - cookie is signed with different secret key than is used by server

Reply:
I understood that I should implement test with expired access token, and mock request to keycloak with refresh token
but should I implement test with signing cookie with different key? imho this logic is not part of syncmaster, it is done in `SessionMiddleware`, and if it signed with different secret key it will handle [BadSignature](https://github.com/encode/starlette/blob/7d586f7e39f3e83e8b86636cd41d75e9a879a791/starlette/middleware/sessions.py#L52) error and not add cookie to `request.session`

Reply:
> but should I implement test with signing cookie with different key?

That's just a question, what SyncMaster will do in this case. Does `SessionMiddleware` reset this cookie if it cannot be decrypted?

Reply:
cookie will be kept in `request.cookie` but not in `request.session` and syncmaster retrieves jwt token from `request.session` therefore it will redirect to keycloak

Reply:
added test with expired access_token